core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java

```diff
@@ -100,6 +106,81 @@ public static EnumerableHashJoin create(
         condition, variablesSet, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null
+        || collation == RelCollations.EMPTY
+        || joinType == JoinRelType.FULL
+        || joinType == JoinRelType.RIGHT) {
+      return null;
+    }
+
+    List<Integer> requiredKeys = RelCollations.ordinals(collation);
+    ImmutableBitSet requiredKeySet = ImmutableBitSet.of(requiredKeys);
+    ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
+
+    // EnumerableHashJoin traits passdown shall only pass through collation to left input.
+    // It is because for EnumerableHashJoin always builds hash table on right input,
+    // and only non-hashed side can preserve ordering.
+
+    // Only consider exact key match for now
+    if (requiredKeySet.equals(leftKeySet)) {
```

Comment:
hash join / nested loopjoin are different with MergeJoin. The required collation keys can be any keys other than join keys. 
```sql
select * from foo join bar using (a) order by foo.b;
```
We can still pass down the collation request to HashJoin's left child, because hashjoin doesn't require any collation from its left child.
But for mergejoin, we can't pass down. because the join's parent requires collation on `b`, but the mergejoin requires its children to be sorted on `a`.

Reply:
This is a greta point.